config: allow setting -proxy per network

vasild opened the pull request:
`-proxy=addr:port` specifies the proxy for all networks (except I2P). Previously only the Tor proxy could have been specified separately via `-onion=addr:port`.

Make it possible to specify separately the proxy for IPv4, IPv6, Tor and CJDNS by e.g. `-proxy=addr:port=ipv6`. Or remove the proxy for a given network, e.g. `-proxy=0=cjdns`.

Resolves: https://github.com/bitcoin/bitcoin/issues/24450

vasild committed:
config: allow setting -proxy per network

`-proxy=addr:port` specifies the proxy for all networks (except I2P).
Previously only the Tor proxy could have been specified separately via
`-onion=addr:port`.

Make it possible to specify separately the proxy for IPv4, IPv6, Tor and
CJDNS by e.g. `-proxy=addr:port=ipv6`. Or remove the proxy for a given
network, e.g. `-proxy=0=cjdns`.

Resolves: https://github.com/bitcoin/bitcoin/issues/24450

vasild commented on src/init.cpp:
```diff
@@ -1189,31 +1204,34 @@ bool CheckHostPortOptions(const ArgsManager& args) {
         }
     }
 
-    for ([[maybe_unused]] const auto& [arg, unix] : std::vector<std::pair<std::string, bool>>{
-        // arg name            UNIX socket support
-        {"-i2psam",                 false},
-        {"-onion",                  true},
-        {"-proxy",                  true},
-        {"-rpcbind",                false},
-        {"-torcontrol",             false},
-        {"-whitebind",              false},
-        {"-zmqpubhashblock",        true},
-        {"-zmqpubhashtx",           true},
-        {"-zmqpubrawblock",         true},
-        {"-zmqpubrawtx",            true},
-        {"-zmqpubsequence",         true},
+    for ([[maybe_unused]] const auto& [param_name, unix, suffix_allowed] : std::vector<std::tuple<std::string, bool, bool>>{
+        // arg name          UNIX socket support  =suffix allowed
+        {"-i2psam",          false,               false},
+        {"-onion",           true,                false},
+        {"-proxy",           true,                true},
+        {"-bind",            false,               true},
+        {"-rpcbind",         false,               false},
+        {"-torcontrol",      false,               false},
+        {"-whitebind",       false,               false},
+        {"-zmqpubhashblock", true,                false},
+        {"-zmqpubhashtx",    true,                false},
+        {"-zmqpubrawblock",  true,                false},
+        {"-zmqpubrawtx",     true,                false},
+        {"-zmqpubsequence",  true,                false},
     }) {
```
note: `-bind` was added here. I guess before it was omitted because this loop did not support `=whatever` suffix. Now it does.

caesrcd reviewed (commented):
ACK 6a9f04688f

I have reviewed the code, unit tests, and functional tests. I built the project and ran a node using the `-proxy=0=cjdns` parameter to verify the behavior. Confirmed that CJDNS correctly bypasses the general proxy as intended. Manual testing was performed on Arch Linux, and the node successfully connected over CJDNS directly without using the proxy.

pablomartin4btc reviewed (commented):
If the user sets `-proxy=addr:port=tor` then `-onion=` shouldn't be allowed, no? Otherwise one could set 2 diff proxies for Tor(?).

Also, in [Tor's documentation](https://github.com/bitcoin/bitcoin/tree/master/doc/tor.md):

```
-proxy=ip:port  Set the proxy server. If SOCKS5 is selected (default), this proxy
                server will be used to try to reach .onion addresses as well.
```

So, shouldn't `-proxy=0=cjdns` be the default if `-cjdnsreachable` when `-proxy=addr:port`?

vasild committed:
doc: update tor.md to mention the new -proxy=addr:port=tor

Also change the formatting of `tor.md` to have more horizonal space for
the text.

Co-authored-by: Matthew Zipkin <pinheadmz@gmail.com>

vasild commented on src/init.cpp:
```diff
@@ -1572,30 +1590,64 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+            if (IsUnixSocketPath(proxy_str)) {
+                proxy = Proxy{proxy_str, /*tor_stream_isolation=*/proxyRandomize};
+            } else {
+                const std::optional<CService> addr{Lookup(proxy_str, 9050, fNameLookup)};
```
Done.

vasild commented:
`6a9f04688f...2059eaa692`: address suggestions

> If the user sets `-proxy=addr:port=tor` then `-onion=` shouldn't be allowed, no? Otherwise one could set 2 diff proxies for Tor(?).

I think that is ok, assuming the expectation is that later options override earlier ones. For example, without this PR, `-proxy=x -onion=y` is supposed to set the Tor proxy to `y`, overriding the earlier `x`. Updated `tor.md` to mention that.

> So, shouldn't `-proxy=0=cjdns` be the default if `-cjdnsreachable` when `-proxy=addr:port`?

Could be, but I do not like these "automatically set `x` only if `y` and `z` are true, otherwise something else". That would also change the default behavior in a breaking way - for people that use `-proxy` which does support CJDNS it will stop working. Maybe there are no such users, but we can't know for sure.

pablomartin4btc commented:
>> Otherwise one could set 2 diff proxies for Tor(?).

> I think that is ok, assuming the expectation is that later options override earlier ones. For example, without this PR, `-proxy=x -onion=y` is supposed to set the Tor proxy to `y`, overriding the earlier `x`. Updated `tor.md` to mention that.

Ok.
 
> > So, shouldn't `-proxy=0=cjdns` be the default if `-cjdnsreachable` when `-proxy=addr:port`?
> 
> Could be, but I do not like these "automatically set `x` only if `y` and `z` are true, otherwise something else". That would also change the default behavior in a breaking way - for people that use `-proxy` which does support CJDNS it will stop working. Maybe there are no such users, but we can't know for sure.

Ok, but since the [previous statement](https://github.com/bitcoin/bitcoin/pull/32425#pullrequestreview-2818995555) (`-proxy=`...) from [Tor's doc](https://github.com/bitcoin/bitcoin/tree/master/doc/tor.md) combined with:

`-onlynet=onion  Make automatic outbound connections only to .onion addresses.`

Maybe we should give a warning at least (and perhaps suggesting: `-proxy=0=cjdns`) because that (`-proxy=ip[:<port>] -onlynet=onion -cjdnsreachable`) would make CJDNS not to work (?).

I think a more long term solution could be to add the possibility of having a separate proxy for CJDNS, as suggested on the [issue](https://github.com/bitcoin/bitcoin/issues/24450) by other commenters there.

caesrcd reviewed (commented):
ACK 2059eaa692ff614c41dee50f4bc2d2946e5d42af

I have tested the code.

pinheadmz commented on doc/tor.md:
```diff
@@ -32,33 +32,43 @@ You can use the `getnodeaddresses` RPC to fetch a number of onion peers known to
-                    Inbound and manual connections are not affected by this option.
-                    It can be specified multiple times to allow multiple networks,
-                    e.g. onlynet=onion, onlynet=i2p, onlynet=cjdns.
+    -proxy=ip[:port]
+        Set the proxy server. It will be used to try to reach .onion addresses
+        as well. You need to use -noonion or -onion=0 to explicitly disable
+        outbound access to onion services.
+
+    -proxy=ip[:port]=tor
+    or
+    -onion=ip[:port]
+        Set the proxy server for reaching .onion addresses. You do not need to
+        set this if it's the same as the generic -proxy. You can use -onion=0 to
+        explicitly disable access to onion services.
+        ------------------------------------------------------------------------
+        Note: Only the -proxy option sets the proxy for DNS requests; with
+        -onion they will not route over Tor, so use -proxy if you have privacy
+        concerns.
+        ------------------------------------------------------------------------
+
+    If -proxy or -onion is specified multiple times, later occurences override
+    earlier ones and command line overrides the config file.
```
Since you're touching this doc, you could also add a note about UNIX sockets which I should've added in #27375 ;-)

e.g.
```
UNIX domain sockets may be used for proxy connections. Set `-onion` or `-proxy`
to the local socket path with the prefix `unix:` (e.g. `-onion=unix:/home/me/torsocket`).
```

vasild replied:
Done.

pinheadmz reviewed (approved):
ACK 2059eaa692ff614c41dee50f4bc2d2946e5d42af

Built and ran all tests on macos/arm64. Reviewed both commits and left a few nits.
Played with the feature on mainnet, and connected to all networks:

```
         ipv4    ipv6   onion   cjdns   total   block  manual
in          0       0       0       0       0
out         1       1       2       7      11       0       7
total       1       1       2       7      11
```

Even used tor's ipv4 and unix sockets for separate proxies ;-)

`bcd -debug=tor -debug=net -proxy=unix:/Users/matthewzipkin/tor/sock -proxy=127.0.0.1:9050=ipv6  -cjdnsreachable -proxy=0=cjdns`


<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 2059eaa692ff614c41dee50f4bc2d2946e5d42af
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmgiWz0ACgkQ5+KYS2KJ
yTp0Ag/+JZCDfs0ILvoCRuuUN74qNh1vlvS9eqie3d32quX4wjYE8ZQC4Ov18xlv
8nOCKzDu369bixp1N6HlQXqa41w0x0KbAshMHWsHwolKr5MumQTfOp6mcnMaD1J4
zbS1Jbyigs7x+ukeVLW22CH9I64GUGQuizmzuWA2/cR0ehFtkLsKyG/5HCSeQU4H
/VzzEGcYNeqKKroLKKswqbkcVSrzhz8I1P7ZOEMhYQ6kDTZ5YVKGhbqcb8n2S8nJ
Tr5GY72gx8WY+LwgLM3nuesus8jYSVKSwleWCGxm2uqKdAkLcBQajBnJy1kgIYO0
ieAMypz9OF8zYls6VvhuTZseTvMWIM0ai37/1YUS8R3Dk3c9habUhv2GoxEpHmZb
2vcJ133qtVRNj9ZWAy987lmXKz1m3imrH+SGGUdG0etaHjs6qz0JnFxv11/xTzaB
SBFc8nalkIG/RZqy72i3WskMdOxkywkVx2fzN5RYZdpxSb2krlaOVln3On60H8iW
Ewnk+MioS0s/1mvp+u31YG5GzRhqO4mCtQMG5IE2W+k58q1gMvq6uQtasgWGu9LX
ZihaLz2cxeHxCqQFaFP1rL54IB/2MWXWXOBCEp24JQT/LXAfq1sacikd1PzaI/T8
tVQcYWs1W3ZORKjqMhWaPEpXurz5mstul/zV/BkRcod6E9XgJPw=
=65/9
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>

vasild commented on src/init.cpp:
```diff
@@ -1572,30 +1590,64 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+            if (IsUnixSocketPath(proxy_str)) {
+                proxy = Proxy{proxy_str, /*tor_stream_isolation=*/proxyRandomize};
+            } else {
+                const std::optional<CService> addr{Lookup(proxy_str, DEFAULT_TOR_SOCKS_PORT, fNameLookup)};
```
Yes, that is the same in `master` so no change in behavior by this PR. In master it is using a magic number `9050` whereas here it uses the constant `DEFAULT_TOR_SOCKS_PORT`.

`-proxy=127.0.0.1=cjdns` would be interpreted as `-proxy=127.0.0.1:9050=cjdns` which might be the correct config, or if not then the user could use `-proxy=127.0.0.1:1234=cjdns` to specify another port.

Personally, I would favor a more dumb implementation that does not do such automations and guessings on behalf of the user. Such an implementation would require the port to always be provided and refuse to start if not.

vasild commented:
`2059eaa692...e98c51fcce`: address suggestions

pinheadmz reviewed (approved):
ACK e98c51fcce9ae3f441a416cab32a5c85756c6c64

Trivial changes since last ACK. Built and tested again on mainnet, this time added i2p as well, connected to peers on every network we support. Tried various proxy settings.

The code change is small and mostly just contained in `init.cpp`. We used to call `SetProxy()` on every network type with the same proxy endpoint, this new code parses more fine-grained config options and sets separate proxies based on the options.

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK e98c51fcce9ae3f441a416cab32a5c85756c6c64
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmgjTV8ACgkQ5+KYS2KJ
yToKAw/6AgyTgGUOnk+VxgZGTTFufgNES5uvasHJtpkvcNYf9boWGpVQXBxf9ePc
BR7ly5C1+Tmvv6X36FwLE+mhNbXHIK178lfWFs+CMMeeuEThwR7LapGSbdAZGa+l
RxXxAx49/Ns5wMnMXPS+7mDp3CKAk1HRm1u1AN5k99RcyXdG5QUqfCb5QctckTUZ
jS17RuzONdJxW34Xd+p2nfsbKYUBy0YQPfLyxjy4e3KaStmpQTtp0HIKampli4t4
GhJiD+Yxj3ohJHqMS7so7jFo6nXqYw9k8wNUdaZTySBwMk2vDzRYesoBuXQZqM+j
xV9KReg1zxbtJTtW4cC38Rr3qRJJm4AKUNQmybs5aDfT07UAe2Ocd+b2lX97Pg4L
EyFCneylgs9mY+XH6UJpbOMihGWjAzZ63dRGrnMEFwZuNOmSBCSqPgQNjqSyBnU1
Cwj8liKCzr+mYwgWSA1M57BJGc0ZYeIPOEi8NEBzwejtQN6ljESmtWXh3kgKcXyb
VE2e3KJvCTTtdjI8r6nk2cNGKDMwgzeJp1DAAMEj2xOT7rI0qYc1tdr4HUGhhl0w
TiCeCenZhgPyTocW0xVFTzAmvhm+WCHp6Epyucjo0hnK4esw965z2N3B+4tmCfMY
B6VJFHuUqaerSIlvFzI8QpJsP9UBUUe2VS+XXhBv1VnlNuGewDM=
=PiAP
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>

caesrcd reviewed (commented):
reACK e98c51fcce

danielabrozzoni reviewed (commented):
Code Review ACK e98c51fcce9ae3f441a416cab32a5c85756c6c64

I have reviewed the code, and it looks good to me.
I run the tests locally and did some very limited manual testing, mimicking the functional test behavior (passing in `-proxy=...=network` and checking the `getnetworkinfo` output)

luke-jr commented on src/init.cpp:
```diff
@@ -1569,33 +1587,66 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
-        onion_proxy = addrProxy;
+        if (net_str.empty()) { // For all networks.
+            ipv4_proxy = ipv6_proxy = name_proxy = cjdns_proxy = onion_proxy = proxy;
+        } else if (net_str == "ipv4") {
+            ipv4_proxy = name_proxy = proxy;
+        } else if (net_str == "ipv6") {
+            ipv6_proxy = name_proxy = proxy;
```
I'm not sure this logic is ideal. If you set an all-network proxy, but then disable it for IPv6, you might be expecting name resolution to use the proxy. Or maybe not.

vasild replied:
That would require some convoluted juggling with the parameters. Like:

1. `-proxy=addr:port=ipv6` sets the "ipv6" and "name" proxy, so that just a single `-proxy=1.2.3.4:5678=ipv6` works as expected. However:
2. `-proxy=0=ipv6` removes _only_ the "ipv6" proxy and not the "name" proxy, so that what you describe works: `-proxy=addr:port -proxy=0=ipv6`.

Another option would be to provide a separate name proxy, e.g. `-proxy=addr:port=name` or `-proxy=addr:port=dns`. But then users will probably forget to set it an be surprised that `-proxy=addr:port=ipv4` does DNS requests without of the proxy. Then it would be tempting for us to automatically set the "name" proxy from the "ipv4" or "ipv6" proxy which is another convoluted automation.

I prefer to keep it simple as it is now, without having mismatching enable/disable logic (1. and 2. above) or automatically setting parameters based on other parameters.

luke-jr commented on src/init.cpp:
```diff
@@ -1569,33 +1587,66 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    Proxy cjdns_proxy;
+    for (const std::string& param_value : args.GetArgs("-proxy")) {
+        const auto eq_pos{param_value.rfind('=')};
+        const std::string proxy_str{param_value.substr(0, eq_pos)}; // e.g. 127.0.0.1:9050=ipv4 -> 127.0.0.1:9050
```
string_view?

vasild replied:
That would make the string_view point to the temporary returned by `substr()`. That temporary is destroyed after the end of the statement, so the string_view remains with a dangling pointer. Like in the example in https://en.cppreference.com/w/cpp/string/basic_string/operator_basic_string_view which says "// Bad: holds a dangling pointer".

Or a small standalone program to demonstrate:

```cpp
int main(int, char**)
{
    std::string s{"abcd"};
    std::string_view sv{s.substr(2)};
    std::cout << sv << "\n";
    return 0;
}
```
Compile with `clang -fsanitize=address` and run:
```
==13945==ERROR: AddressSanitizer: stack-use-after-scope on address ...
```

luke-jr reviewed (commented):
Probably an improvement as-is, but might want to think through the name proxy logic

luke-jr commented on doc/tor.md:
```diff
@@ -32,33 +32,53 @@ You can use the `getnodeaddresses` RPC to fetch a number of onion peers known to
+        service provider.
+        ------------------------------------------------------------------------
+
+If -proxy or -onion is specified multiple times, later occurences override
+earlier ones and command line overrides the config file. UNIX domain sockets may
+be used for proxy connections. Set `-onion` or `-proxy` to the local socket path
+with the prefix `unix:` (e.g. `-onion=unix:/home/me/torsocket`).
+
```
Did you want to indent this?

vasild replied:
No, because when indented then it is rendered as a code block.

OK:
---

## 1. Run Bitcoin Core behind a Tor proxy

The first step is running Bitcoin Core behind a Tor proxy. This will already anonymize all
outgoing connections, but more is possible.

    -proxy=ip[:port]
        Set the proxy server. It will be used to try to reach .onion addresses
        as well. You need to use -noonion or -onion=0 to explicitly disable
        outbound access to onion services.

    -proxy=ip[:port]=tor
    or
    -onion=ip[:port]
        Set the proxy server for reaching .onion addresses. You do not need to
        set this if it's the same as the generic -proxy. You can use -onion=0 to
        explicitly disable access to onion services.
        ------------------------------------------------------------------------
        Note: The proxy for DNS requests is taken from
        -proxy=addr:port or
        -proxy=addr:port=ipv4 or
        -proxy=addr:port=ipv6
        (last one if multiple options are given). It is not taken from
        -proxy=addr:port=tor or
        -onion=addr:port.
        If no proxy for DNS requests is configured, then they will be done using
        the functions provided by the operating system, most likely resulting in
        them being done over the clearnet to the DNS servers of the internet
        service provider.
        ------------------------------------------------------------------------

If -proxy or -onion is specified multiple times, later occurences override
earlier ones and command line overrides the config file. UNIX domain sockets may
be used for proxy connections. Set `-onion` or `-proxy` to the local socket path
with the prefix `unix:` (e.g. `-onion=unix:/home/me/torsocket`).

    -listen
        When using -proxy, listening is disabled by default. If you want to                                                                                             
        manually configure an onion service (see section 3), you'll need to
        enable it explicitly.

    -connect=X
    -addnode=X
    -seednode=X
        When behind a Tor proxy, you can specify .onion addresses instead of IP
        addresses or hostnames in these parameters. Such addresses can also be
        exchanged with other P2P nodes.

    -onlynet=onion
        Make automatic outbound connections only to .onion addresses. Inbound
        and manual connections are not affected by this option. It can be
        specified multiple times to allow multiple networks, e.g. onlynet=onion,
        onlynet=i2p, onlynet=cjdns.

In a typical situation, this suffices to run behind a Tor proxy:

    ./bitcoind -proxy=127.0.0.1:9050

NOT OK:
---

## 1. Run Bitcoin Core behind a Tor proxy

The first step is running Bitcoin Core behind a Tor proxy. This will already anonymize all
outgoing connections, but more is possible.

    -proxy=ip[:port]
        Set the proxy server. It will be used to try to reach .onion addresses
        as well. You need to use -noonion or -onion=0 to explicitly disable
        outbound access to onion services.

    -proxy=ip[:port]=tor
    or
    -onion=ip[:port]
        Set the proxy server for reaching .onion addresses. You do not need to
        set this if it's the same as the generic -proxy. You can use -onion=0 to
        explicitly disable access to onion services.
        ------------------------------------------------------------------------
        Note: The proxy for DNS requests is taken from
        -proxy=addr:port or
        -proxy=addr:port=ipv4 or
        -proxy=addr:port=ipv6
        (last one if multiple options are given). It is not taken from
        -proxy=addr:port=tor or
        -onion=addr:port.
        If no proxy for DNS requests is configured, then they will be done using
        the functions provided by the operating system, most likely resulting in
        them being done over the clearnet to the DNS servers of the internet
        service provider.
        ------------------------------------------------------------------------

    If -proxy or -onion is specified multiple times, later occurences override
    earlier ones and command line overrides the config file. UNIX domain sockets may
    be used for proxy connections. Set `-onion` or `-proxy` to the local socket path
    with the prefix `unix:` (e.g. `-onion=unix:/home/me/torsocket`).

    -listen
        When using -proxy, listening is disabled by default. If you want to                                                                                             
        manually configure an onion service (see section 3), you'll need to
        enable it explicitly.

    -connect=X
    -addnode=X
    -seednode=X
        When behind a Tor proxy, you can specify .onion addresses instead of IP
        addresses or hostnames in these parameters. Such addresses can also be
        exchanged with other P2P nodes.

    -onlynet=onion
        Make automatic outbound connections only to .onion addresses. Inbound
        and manual connections are not affected by this option. It can be
        specified multiple times to allow multiple networks, e.g. onlynet=onion,
        onlynet=i2p, onlynet=cjdns.

In a typical situation, this suffices to run behind a Tor proxy:

    ./bitcoind -proxy=127.0.0.1:9050

vasild commented on src/init.cpp:
```diff
@@ -1569,33 +1587,66 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+            return InitError(strprintf(_("Unrecognized network in -proxy='%s': '%s'"), param_value, net_str));
+        }
+    }
+    if (ipv4_proxy.IsValid()) {
```
Right. But I prefer to be explicit rather than feed `SetProxy()` an invalid value and rely on it doing the "right thing". It is not documented what `SetProxy()` would do with an invalid input, so I prefer to have these checks.

1440000bytes reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/32425/commits/e98c51fcce9ae3f441a416cab32a5c85756c6c64

<details>
<summary>Signature</summary>

<pre>
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK e98c51fcce9ae3f441a416cab32a5c85756c6c64
-----BEGIN PGP SIGNATURE-----

iHUEARYKAB0WIQSSUwYqT5LWNFkIXKYtIwUgISpZAQUCaCuzGQAKCRAtIwUgISpZ
AVt3AP98Co6T1/S4RqaIWypoL4NI1B1PhTPdRdFdvPP0sHWxIgD/SlBxLKNBIehf
tvLSnyIV9cvCBRefydIJCTf+dS06tAw=
=D/ae
-----END PGP SIGNATURE-----
</pre>
</details>

achow101 commented on src/init.cpp:
```diff
@@ -1569,33 +1587,66 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    Proxy name_proxy;
+    Proxy cjdns_proxy;
+    for (const std::string& param_value : args.GetArgs("-proxy")) {
+        const auto eq_pos{param_value.rfind('=')};
+        const std::string proxy_str{param_value.substr(0, eq_pos)}; // e.g. 127.0.0.1:9050=ipv4 -> 127.0.0.1:9050
```
In ca5781e23a8f299ff4f143d2355218f551e65944 "config: allow setting -proxy per network"

nit: `Split()`? (from `src/util/string.h`)

achow101 commented on src/init.cpp:
```diff
@@ -1189,31 +1204,34 @@ bool CheckHostPortOptions(const ArgsManager& args) {
+        {"-zmqpubsequence",  true,                false},
     }) {
-        for (const std::string& socket_addr : args.GetArgs(arg)) {
+        for (const std::string& param_value : args.GetArgs(param_name)) {
+            const std::string param_value_hostport{
+                suffix_allowed ? param_value.substr(0, param_value.rfind('=')) : param_value};
```
In ca5781e23a8f299ff4f143d2355218f551e65944 "config: allow setting -proxy per network"

nit: I think using `Split()` would make this easier to read

glozow commented:
Needs release note?

vasild commented:
> Needs release note?

Right. Done in https://github.com/bitcoin/bitcoin/pull/32727
